com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/verifier/JarVerifier.java

```diff
@@ -0,0 +1,193 @@
+public class JarVerifier implements Verifier {
+
+    private static final String META_INF = "META-INF";
+    private static final String CN = "CN";
+    private static final int BUFFER_SIZE = 8 * 1024;
+
+    @Override
+    public void verify(List<X509Certificate> targetCertificates, String certificateCN, URL jarURL) {
+        try {
+            JarFile jarFile = retrieveJarFile(jarURL);
+            List<JarEntry> jarEntries = getJarEntries(jarFile);
+            verifyJarEntries(targetCertificates, jarEntries, certificateCN);
+        } catch (IOException e) {
+            throw new ResourceAccessException(Messages.CERTIFICATE_VERIFICATION_HAS_FAILED, e);
+        }
+    }
+
+    private JarFile retrieveJarFile(URL jarURL) throws IOException {
+        jarURL = new URL("jar:" + jarURL.toString() + "!/");
+        JarURLConnection connection = (JarURLConnection) jarURL.openConnection();
+        return connection.getJarFile();
+    }
+
+    private List<JarEntry> getJarEntries(JarFile jarFile) throws IOException {
+        List<JarEntry> jarEntries = new ArrayList<>();
+        Enumeration<JarEntry> entries = jarFile.entries();
+        while (entries.hasMoreElements()) {
+            readJarEntry(jarFile, jarEntries, entries);
+        }
+        return jarEntries;
+    }
+
+    private void readJarEntry(JarFile jarFile, List<JarEntry> jarEntries, Enumeration<JarEntry> entries) throws IOException {
+        JarEntry jarEntry = entries.nextElement();
+        validateJarEntryPath(jarFile, jarEntry);
+        if (jarEntry.isDirectory()) {
+            return;
+        }
+        jarEntries.add(jarEntry);
+        readStream(jarFile, jarEntry);
+    }
+
+    private void validateJarEntryPath(JarFile jarFile, JarEntry jarEntry) throws IOException {
+        Path currentRelativePath = Paths.get("");
+        File jarEntryFile = new File(jarEntry.getName());
+        if (!jarEntryFile.getCanonicalPath()
+                         .startsWith(currentRelativePath.toAbsolutePath() + File.separator)) {
+            throw new IllegalAccessError(MessageFormat.format(Messages.ENTRY_IS_TRYING_TO_LEAVE_TARGET_DIR, jarFile.getName()));
+        }
+    }
+
+    private void readStream(JarFile jarFile, JarEntry jarEntry) throws IOException {
+        byte[] buffer = new byte[BUFFER_SIZE];
+
+        try (InputStream inputStream = jarFile.getInputStream(jarEntry)) {
+            while (inputStream.read(buffer, 0, buffer.length) != -1) {
+                // Just read. This will throw a SecurityException if a signature/digest check fails.
+            }
+        }
+    }
+
+    private void verifyJarEntries(List<X509Certificate> targetCerts, List<JarEntry> jarEntries, String certificateCN) {
+        for (JarEntry jarEntry : jarEntries) {
+            Certificate[] entryCertificates = jarEntry.getCertificates();
+            if (entryCertificates == null || entryCertificates.length == 0) {
+                checkJarEntryName(jarEntry);
+                continue;
```

Comment:
Invert the condition to avoid this continue

Reply:
If I invert the condition and remove the continue one or both of the actions will be executed, I want to check the jar entry name or verify jar entry is signed (One of them). I use it instead of else.